src/AutoHashEqualsCached.jl

```diff
@@ -0,0 +1,212 @@
+  result in infinite recursion.
+"""
+function _show_default_auto_hash_equals_cached(io::IO, @nospecialize(x))
+    t = typeof(x)
+    show(io, Base.inferencebarrier(t)::DataType)
+    print(io, '(')
+    recur_io = IOContext(io, Pair{Symbol,Any}(:SHOWN_SET, x),
+                         Pair{Symbol,Any}(:typeinfo, Any))
+    if !Base.show_circular(io, x)
+        for i in 1:nfields(x)
+            f = fieldname(t, i)
+            if (f == :_cached_hash)
+                continue
+            elseif i > 1
+                print(io, ", ")
+            end
+            if !isdefined(x, f)
+                print(io, Base.undef_ref_str)
+            else
+                show(recur_io, getfield(x, i))
+            end
+        end
+    end
+    print(io,')')
+end
+
+"""
+Find the first struct declaration buried in the Expr.
+"""
+get_struct_decl(typ) = nothing
+function get_struct_decl(typ::Expr)
+    if typ.head == :struct
+        return typ
+    elseif typ.head == :macrocall
+        return get_struct_decl(typ.args[3])
+    elseif typ.head == :block
+        # get the first struct decl in the block
+        for x in typ.args
+            if x.head == :macrocall || x.head == :struct || x.head == :block
+                result = get_struct_decl(x)
+                if !isnothing(result)
+                    return result
+                end
+            end
+        end
+    end
+
+    error("macro @auto_hash_equals_cached should only be applied to a struct")
+end
+
+unpack_name(node) = node
+function unpack_name(node::Expr)
+    if node.head == :macrocall
+        unpack_name(node.args[3])
+    elseif node.head in (:(<:), :(::))
+        unpack_name(node.args[1])
+    else
+        node
+    end
+end
+
+unpack_type_name(n::Symbol) = (n, n, nothing)
+function unpack_type_name(n::Expr)
+    if n.head == :curly
+        type_name = n.args[1]
+        type_name isa Symbol ||
+            error("macro @auto_hash_equals_cached applied to type with invalid signature: $type_name")
+        where_list = n.args[2:length(n.args)]
+        type_params = map(unpack_name, where_list)
+        full_type_name = Expr(:curly, type_name, type_params...)
+        (type_name, full_type_name, where_list)
+    elseif n.head == :(<:)
+        unpack_type_name(n.args[1])
+    else
+        error("macro @auto_hash_equals_cached applied to type with unexpected signature: $n")
+    end
+end
+
+function get_fields(struct_decl::Expr)
+    member_names = Vector{Symbol}()
+    member_decls = Vector()
+
+    add_field(b) = nothing
+    function add_field(b::Symbol)
+        push!(member_names, b)
+        push!(member_decls, b)
+    end
+    function add_field(b::Expr)
+        if b.head == :block
+            add_fields(field)
+        elseif b.head == :const
+            add_field(b.args[1])
+        elseif b.head == :(::) && b.args[1] isa Symbol
+            push!(member_names, b.args[1])
+            push!(member_decls, b)
```

Comment:
should a case be added in this function to error on internal constructors in the struct definition?

Reply:
Yes; done.